crates/control/src/services/builds_root.rs

```diff
@@ -0,0 +1,344 @@
+use crate::config;
+use crate::error::SubprocessError;
+use async_trait::async_trait;
+use rusqlite::{Connection, OpenFlags};
+use std::collections::BTreeMap;
+use std::fmt::Debug;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+use tokio::sync::Mutex;
+
+mod gcs;
+
+/// Allows uploading builds to the builds root.
+#[derive(Debug, Clone)]
+pub struct PutBuilds(Arc<dyn BuildsRootService>);
+
+impl PutBuilds {
+    /// Uploads the given `build_db` to the builds root so that it is accessible to the data plane.
+    async fn put_build(&self, build_id: &str, build_db: &Path) -> Result<(), BuildsRootError> {
+        self.0.put_build(build_id, build_db).await
+    }
+}
+
+/// A sqlite connection to a build database that has been fetched from the builds root.
+/// This type implements `Deref<Target=rusqlite::Connection>`, so you can use it just like you
+/// would a normal sqlite connection. This is an inert wrapper type currently, and exists to allow
+/// for caching in the future.
+#[derive(Debug)]
+pub struct BuildDBRef {
+    connection: Connection,
+}
+
+impl std::ops::Deref for BuildDBRef {
+    type Target = Connection;
+
+    fn deref(&self) -> &Self::Target {
+        &self.connection
+    }
+}
+
+/// Allows querying build databases by build id. Builds will be fetched automatically from the
+/// builds root, if required.
+#[derive(Debug, Clone)]
+pub struct FetchBuilds {
+    root: Arc<dyn BuildsRootService>,
+    local_builds: Arc<Mutex<BTreeMap<String, LocalBuildState>>>,
```

Comment:
I don't think it's worth it, because you don't know whether or not you need to write until after you read to see if the build is already present. So a `RwLock` would require two separate lock operations, and you'd still have to re-check to see if the build was present after acquiring the write lock.

Reply:
I think I wrote this comment before fully understanding the caching scheme.